sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs

```diff
@@ -286,7 +286,7 @@ private static void CloseLink(RequestResponseAmqpLink link)
             CancellationToken cancellationToken)
         {
             var link = default(ReceivingAmqpLink);
-            var receivedMessages = new List<ServiceBusReceivedMessage>();
+            var receivedMessages = new List<ServiceBusReceivedMessage>(maxMessages);
```

Comment:
Would this overallocate for cases where the connection is not saturated?

Reply:
Good point, it would, e.g. if someone is requesting 1000 messages but there are only a few returned.

Reply:
This would actually be a common scenario with the Service Bus extensions where I believe the default message count is quite high.

Reply:
Well but the thing is we this also overallocates for cases when we pass 1 for single message receives and over allocates for cases the there are more messages than the default size because the capacity will be grown. So essentially because you are specifying a number of messages to receive you are expecting a certain amount of messages, (which is the memory you will need eventually) that was the tradeoff I was coming from because in essence my thought was the right value is the amount of messages you are willing to fetch.

In fact this is the changes that also got into Amqp libs so if we would want to be extrem we could also directly expose that list of nonempty receives via switch case and wouldn't even need to create a new one

For reference https://github.com/Azure/azure-amqp/pull/177

Reply:
This gets thrown off badly when prefetch is enabled.   The AMQP library does not wait for the requested number of messages in that case; once any message is available in the prefetch queue, it will return whatever happens to be available irrespective of the requested message count or any timeout.    

I know that prefetch use less common in Service Bus than elsewhere, so I'm not sure how impactful the consideration is for this change, but I thought it worth mentioning.

Reply:
I can send in an alternative. How about I isolate this one as a dedicated PR? https://github.com/Azure/azure-sdk-for-net/pull/20550#discussion_r617528248 and send in a proposal to rely on the list of AMQP lib?

Reply:
Never mind, I forgot the fact that we convert the messages.

Reply:
@JoshLove-msft @jsquire @pakrym what do you think about this? https://github.com/Azure/azure-sdk-for-net/pull/20571 Too cheeky?